server/cache/two_queue.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+const (
+
+	// defaultRecentRatio is the ratio of the 2Q cache dedicated
+	// to recently added entries that have only been accessed once.
+	defaultRecentRatio float64 = 0.25
+
+	// defaultGhostRatio is the default ratio of ghost
+	// entries kept to track entries recently evicted
+	defaultGhostRatio float64 = 0.50
+)
+
+// TwoQueue is a fixed size 2Q cache.
+// 2Q is an enhancement over the standard LRU cache
+// in that it tracks both frequently and recently used
+// entries separately. This avoids a burst in access to new
+// entries from evicting frequently used entries. It adds some
+// additional tracking overhead to the standard LRU cache, and is
+// computationally about 2x the cost, and adds some metadata over
+// head. The ARCCache is similar, but does not require setting any
+// parameters.
+// TwoQueue implements is reference by https://github.com/hashicorp/golang-lru/blob/master/2q.go
+type TwoQueue struct {
```

Comment:
why not using hashicorp/golang-lru directly?

Reply:
Because our cache's key is uint64 and I want to use our LRU implements instead  hashicorp/golang-lru